app/src/org/commcare/tasks/FormRecordToFileTask.java

```diff
@@ -0,0 +1,278 @@
+        results = null;
+    }
+
+    private static final String[] SUPPORTED_FILE_EXTS = {".xml", ".jpg", ".3gpp", ".3gp"};
+
+    /**
+     * Turn a FormRecord folder from storage into a standard file representation in our file system.n
+     */
+    private long getFileInstanceFromStorage(File formRecordFolder, SecretKeySpec decryptionKey)
+            throws FileNotFoundException, SessionUnavailableException {
+        File[] files = formRecordFolder.listFiles();
+        Logger.log(TAG, "Trying to get instance with: " + files.length + " files.");
+
+        File myDir = new File(storedFormDirectory, formRecordFolder.getName());
+        myDir.mkdirs();
+
+        if (files == null) {
+            //make sure external storage is available to begin with.
+            String state = Environment.getExternalStorageState();
+            if (!Environment.MEDIA_MOUNTED.equals(state)) {
+                //If so, just bail as if the user had logged out.
+                throw new SessionUnavailableException("External Storage Removed");
+            } else {
+                throw new FileNotFoundException("No directory found at: " + formRecordFolder.getAbsoluteFile());
+            }
+        }
+
+        //If we're listening, figure out how much (roughly) we have to send
+        long bytes = 0;
+        for (File file : files) {
+            //Make sure we'll be sending it
+            boolean supported = false;
+            for (String ext : SUPPORTED_FILE_EXTS) {
+                if (file.getName().endsWith(ext)) {
+                    supported = true;
+                    break;
+                }
+            }
+            if (!supported) {
+                continue;
+            }
+            bytes += file.length();
+        }
+
+        Log.d(TAG, "Storing " + bytes + " form bytes");
+
+        final Cipher decryptCipher = FormUploadUtil.getDecryptCipher(decryptionKey);
+
+        for (File file : files) {
+            // This is not the ideal long term solution for determining whether we need decryption, but works
+            if (file.getName().endsWith(".xml")) {
+                try {
+                    FileUtil.copyFile(file, new File(myDir, file.getName()), decryptCipher, null);
+                } catch (IOException ie) {
+                    Log.d(TAG, "Copying file: " + file.getName() + " failed with: " + ie.getMessage());
+                    publishProgress(("File writing failed: " + ie.getMessage()));
+                    return FormUploadUtil.FAILURE;
+                }
+            } else {
+                try {
+                    FileUtil.copyFile(file, new File(myDir, file.getName()));
+                } catch (IOException ie) {
+                    Log.d(TAG, "Copying file: " + file.getName() + " failed with: " + ie.getMessage());
+                    publishProgress(("File writing failed: " + ie.getMessage()));
+                    return FormUploadUtil.FAILURE;
+                }
+            }
+        }

```

Comment:
Lifting this `for`-loop into a helper method that throws an `IOException` will help remove the duplicated error handling present here.